src/circuit/v2/util.ts

```diff
@@ -0,0 +1,100 @@
+import { abortableDuplex, AbortError } from 'abortable-iterator'
+import type { Duplex, Sink, Source } from 'it-stream-types'
+import { TimeoutController } from 'timeout-abort-controller'
+import { Uint8ArrayList } from 'uint8arraylist'
+import type { Limit } from './pb/index.js'
+import { logger } from '@libp2p/logger'
+
+const log = logger('libp2p:circuit:v2:util')
+
+type DuplexStream = Duplex<Uint8ArrayList, Uint8ArrayList | Uint8Array>
+const doRelay = (src: DuplexStream, dst: DuplexStream) => {
+  queueMicrotask(() => {
```

Comment:
These are spawned as microtasks to allow for synchronous sources:

consider `dst.sink(src.source); src.sink(dst.source)`. `dst.sink` will block until `src.source` completes if `src.source` is synchronous.